### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by arvidn**
This patch pulls out ip_set and node_buffer into their own headers in order to add unit tests for them. It also adds catch.hpp to the tests directory and updates the existing tests to use it. It adds a unit test for ip_set and node_buffer. While adding tests I discovered two bugs that I fixed. Since I fixed them and moved the code they were in in a single PR, you can only see the fix by looking at the second commit. I'll point it out in a comment though.

### CODE REVIEW

File: node_buffer.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+#pragma once
+
+#include <array>
+#include <algorithm>
+#include "ip_set.hpp"
+#include "mapped_file.hpp" // for mapped_vector
+
+template <typename Address>
+struct node_buffer
```
Comment (arvidn) [inline node_buffer.hpp]: node_buffer_t was renamed just node_buffer

File: node_buffer.hpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+{
+	using address_type = Address;
+	using steady_clock = std::chrono::steady_clock;
+	using node_id_type = std::array<char, 20>;
+
+	// this is the type of each node entry
+	// in the circular buffer
+	struct node_entry_t
```
Comment (arvidn) [inline node_buffer.hpp]: and the buffer entry was made an inner class

File: node_id_type.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (arvidn) [inline node_id_type.hpp]: weird, I meant to delete this file from the commit.

File: node_buffer.hpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	node_buffer(char const* filename, size_t const capacity)
+		: m_current_max_size(std::min(size_t(99), capacity))
+		, m_capacity(capacity)
+		, m_last_write_loop(steady_clock::now())
+		, m_buffer(filename, capacity)
+	{
+		// if we picked up nodes from the previous session, we need to restore our
+		// set of IPs too (used to disallow duplicates) as well as making sure the
+		// current_max_size and write cursors are set correctly
+		if (m_buffer.size() > 0) {
```
Comment (arvidn) [inline node_buffer.hpp]: this logic is new. Previously just the buffer was restored, but the m_ips unordered_set was not rebuilt, and hence asserts would fire when trying to work with an existing file

File: mapped_file.hpp [hunk 5]
```diff
@@ -88,7 +88,7 @@ struct mapped_vector
 {
 	// the header must be large enough to make the first element still be
 	// correctly aligned
-	static constexpr size_t header_size = (16 > alignof(T)) ? alignof(T) : 16;
+	static constexpr size_t header_size = (16 < alignof(T)) ? alignof(T) : 16;
```
Comment (arvidn) [inline mapped_file.hpp]: this was an embarrasing typo.

File: node_buffer.hpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+			for (node_entry_t const& e : m_buffer) {
+				m_ips.insert(address_type(e.ip));
+			}
+			m_write_cursor = m_buffer.size();
+			if (m_write_cursor == m_current_max_size) {
+				m_write_cursor = 0;
+			}
+			if (m_current_max_size < m_buffer.size()) {
```
Comment (ssiloti) [inline node_buffer.hpp]: This block needs to be moved above line 63 to make the condition on line 64 work.

Comment (arvidn) [inline node_buffer.hpp]: ah, yes

**Comment by arvidn** [comment]
I squashed the typo-commits, and fixed the node_buffer constructor

**Comment by ssiloti** [comment]
lgtm

**Pull request merged**